### PR TITLE
Cirrus: Simplify log collection commands

### PR DESCRIPTION
**Pull request by cevich**
Signed-off-by: Chris Evich <cevich@redhat.com>

### CODE REVIEW

**Comment by mheon** [comment]
Changes LGTM

File: .cirrus.yml [hunk 1]
```diff
@@ -280,18 +288,16 @@ testing_task:
...
-        failed_df_script: '${DFCMD}'
-        failed_audit_log_script: 'cat /var/log/audit/audit.log || cat /var/log/kern.log || echo "Uh oh, cat audit.log failed"'
-        failed_journalctl_b_script: 'journalctl -b || echo "Uh oh, journalctl -b failed"'
+
+    always:
+        # Carefuling - Don't fail again and miss collecting all data
+        ginkgo_node_logs_script: ${GINKGOLOGCMD}
+        df_script: ${DFCMD}
+        audit_log_script: ${AUDITLOGCMD}
+        journalctl_b_script: {JRNLLOGCMD}
```
Comment (edsantiago) [inline .cirrus.yml]: Missing dollar sign?

Comment (edsantiago) [reply .cirrus.yml]: Also, do you expect there ever to be a need for reusing the three new XXXCMDs? If not, might it be a lot simpler just to specify them as-is in the `always` block? First, it saves someone the time of having to scroll up and find them; second, it's much more readable and less error-prone than ABRVTDCMD (which, given the unreadable abbreviations, is easily subject to misspelling).

Comment (cevich) [reply .cirrus.yml]: oops.

**Comment by cevich** [comment]
**note**: I plan on doing a pass across the entire ``.cirrus.yml``, but need to verify some cirrus-ci YAML parsing behavior.  IIRC, there was also some strange/unexpected interactions between always/on_failure instructions.  Need to instrument some test failures to confirm.

**Comment by edsantiago** [comment]
Ah - yes, I see those other instances now. Thanks.

**Comment by cevich** [comment]
Ugg, yep, cirrus (I think) YAML parser does really odd/nonstandard things with quotes.

```
cat: '||': No such file or directory
cat: echo: No such file or directory
cat: "'Failed": No such file or directory
cat: to: No such file or directory
cat: collect: No such file or directory
cat: ginkgo: No such file or directory
cat: node: No such file or directory
cat: "logs'": No such file or directory
```

**Comment by edsantiago** [comment]
In the removed code, the commands were enclosed in single-quotes:
```
   failed_df_script: '${DFCMD}'
```
In the added code, no quotes. Should you try adding them back?

**Comment by cevich** [comment]
Naa, taking a different approach.  It's the quote-handling that's the problem.  I've wasted time on this topic before, it's not worth the aggravation...

**Comment by cevich** [comment]
@edsantiago does this approach seem reasonable?  IMHO it rings a bit of overkill, but perhaps is good for future growth and maintenance.  The absence of ginkgo node-logs in some tasks is expected (I think) - depends on what the makefile does.

**Comment by edsantiago** [comment]
@cevich yes, this approach makes sense and looks good. But why not take the next step, and just call the script `logcollector` or something, and make it do everything in just one pass? Every single instance I see has four invocations, all of which are duplication, not to mention the complexity of the script itself; why not just:
```
    always:
        log_collector_script: '${WHATEVER}'
```

**Comment by cevich** [comment]
Ack on name change...I always struggle with these.

Reason for duplicate calls is because it splits the output up in the WebUI and in the  backend storage.  Also because Cirrus truncates the output, we want to keep it as short as possible to avoid additional clicks to access it.

File: contrib/cirrus/lib.sh [hunk 2]
```diff
@@ -139,6 +139,11 @@ die() {
     exit ${1:-1}
 }
 
+showrun() {
```
Comment (edsantiago) [inline contrib/cirrus/lib.sh]: Please remove this; it's not used, and it's confusing. (If you ever end up needing/using it in other code, then OK to add back, but please with a note in logcollector.sh indicating that you're overriding it).

File: contrib/cirrus/logcollector.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+req_env_var CIRRUS_WORKING_DIR OS_RELEASE_ID
+
+# Assume there are other log collection commands to follow - Don't
+# let one break another that may be useful, but also keep any
+# actual script-problems fatal so they are noticed right away.
+showrun() {
+    echo '+ '$(printf " %q" "$@")
+    set +e
+    echo '------------------------------------------------------------'
+    "$@"
```
Comment (edsantiago) [inline contrib/cirrus/logcollector.sh]: How about:
```
status=$?
if [[ $status -ne 0 ]]; then
    echo "[ rc = $status -- proceeding anyway ]"
fi
```

**Comment by edsantiago** [comment]
I hate to suggest making this *five* duplicated stanzas... but I find myself just now desperate for the output of `./bin/podman info` on the rootless test; in particular, the `.store` settings. Is there some non-complicated way of adding that?

**Comment by cevich** [comment]
> I find myself just now desperate for the output of

All rooless testing happens from the `special_testing_rootless_task` now, so adding a new script item like `$SCRIPT_BASE/logcollector.sh podman` there would do it.  That also fixes the issue of stanza duplication, one of them would be slightly different :grin: 

> five duplicated stanzas

Actually, now that you mention it, I wonder if  Cirrus's customized parser can handle YAML anchors and aliases?  I'll give that a try, but leave `special_testing_rootless_task` alone in anticipation of your addition.

File: contrib/cirrus/lib.sh (hunk 2 above)
Comment (cevich) [reply contrib/cirrus/lib.sh]: woops, good catch.

File: contrib/cirrus/logcollector.sh (hunk 3 above)
Comment (cevich) [reply contrib/cirrus/logcollector.sh]: good idea.

**Comment by cevich** [comment]
@edsantiago [anchors/aliases appear to work!](https://confluence.atlassian.com/bitbucket/yaml-anchors-960154027.html)  I even left a comment on why rootless testing doesn't use it :smile:

**Comment by edsantiago** [comment]
Excellent! Thank you for looking into it!

File: .cirrus.yml [hunk 4]
```diff
@@ -316,16 +311,16 @@ special_testing_rootless_task:
...
-    journalctl_b_script: 'journalctl -b'
 
     on_failure:
         failed_master_script: '$CIRRUS_WORKING_DIR/$SCRIPT_BASE/notice_master_failure.sh'
-        # Job has already failed, don't fail again and miss collecting data
-        failed_df_script: '${DFCMD}'
-        failed_audit_log_script: 'cat /var/log/audit/audit.log || cat /var/log/kern.log || echo "Uh oh, cat audit.log failed"'
-        failed_journalctl_b_script: 'journalctl -b || echo "Uh oh, journalctl -b failed"'
+
+    always:  # Not using yaml alias in anticipation of specialization
```
Comment (edsantiago) [inline .cirrus.yml]: Feel free to change this one also: I need to think a lot harder about how to get `podman info`, but I do know that it doesn't belong here. Thanks!

Comment (cevich) [reply .cirrus.yml]: ok

**Comment by cevich** [comment]
Thanks @edsantiago for helping me remember about YAML anchor/alias.  I'm happy to see Cirrus-CI's custom parser supports them.  If it weren't my last day before vacation, I'd be tempted to open a PR and apply that trick all over :grin:

**Comment by cevich** [comment]
(this PR is ready to go)

**Comment by rhatdan** [comment]
/approve
You need to remove WIP if you want it to merge.

**Pull request merged**